fix(`@trpc/server`): rearrange the order of context generation in `resolveHTTPResponse` method

maticzav opened the pull request:
This PR fixes an issue where context would be empty in `onError` and `errorFormatter` functions when calling the TRPC server with invalid JSON data.

We encountered this error while testing by calling an endpoint like `http://localhost:3001/trpc/brands.getDefault?input=QUERY`. Because "QUERY" is not a valid JSON, the default `getJsonContentTypeInputs` (i.e. `getInputs`) function would throw. Since context currently only gets generated *after* the input is parsed, it wouldn't be created correctly.

> We use context to communicate through the application the configuration parameters. One of the configurations is the production/development stage which we use to determine whether error should be transformed. Because of the bug, the transformer wouldn't work as expected and essentially leaked information about our file system in the stack trace.

```ts
const t = initTRPC.context<Context>().create({
  transformer: superjson,
  errorFormatter: ({ shape, ctx }) => {
    if (ctx?.config.isProduction) {
      return {
        message: 'Internal server error',
        code: -32603 as any,
        data: {},
      }
    }

    return shape
  },
})
```

I believe this is not intentional and this PR fixes the bug by changing the order of context generation and input parsing. This way `caughtErrorToData` gets the correct context value even if `getInputs` throws on invalid JSON.

Sheraff commented:
@maticzav Could you leave a comment in the code to explain why `ctx` needs to be created before it's used? (explaining the code path that would result in an issue if not created before `getInputs` is called)

Sheraff reviewed (commented):
Could you leave a comment in the code to explain why ctx needs to be created before it's used? (explaining the code path that would result in an issue if not created before getInputs is called)

maticzav commented:
@Sheraff I updated the code and added a comment that seemed in line with the structure of other comments. Let me know if that's what you had in mind! 🙂

Sheraff commented:
> @Sheraff I updated the code and added a comment that seemed in line with the structure of other comments. Let me know if that's what you had in mind! 🙂

That's great, very clear, thanks. And good catch.

juliusmarminge commented on .changeset/twelve-rocks-look.md:
We dont use changesets, can you remove this?

maticzav replied:
Yes!

KATT commented:
FYI, this will be removed in the next major as we have a new `info`-object passed to the `createContext` which contains info about the procedures that will be ran and their inputs
